### PR TITLE
Empty intersection reduction

### LINKED ISSUE #31663: Type 'any' is not assignable to type 'never' with boolean in interface in TS 3.5
**TypeScript Version:**  3.5.1

**Search Terms:**

- Type 'any' is not assignable to type 'never'

**Code**

```ts
interface T {
  str: string;
  bool: boolean;
}

const initState: T = {
  str: 'date',
  bool: true,
};

   let k: keyof T;
   initState[k] = 'test' as any;
// ~~~~~~~~~~~~ Type 'any' is not assignable to type 'never'.ts(2322)
```

This seems to happen specifically with `boolean` properties. If I get rid of the `boolean`-valued property from the interface it works fine. If I have a larger interface and select non-`boolean` properties, it also works fine. But `boolean` and another type produces this error.

This error happens with TS 3.5 but not 3.4 (see playground link).

**Expected behavior:**

No error.

**Actual behavior:**

`Type 'any' is not assignable to type 'never'`

**Playground Link:**  [link][1]

**Related Issues:** 

[1]: https://www.typescriptlang.org/play/#src=interface%20T%20%7B%0D%0A%20%20str%3A%20string%3B%0D%0A%20%20bool%3A%20boolean%3B%0D%0A%7D%0D%0A%0D%0Aconst%20initState%3A%20T%20%3D%20%7B%0D%0A%20%20str%3A%20'date'%2C%0D%0A%20%20bool%3A%20true%2C%0D%0A%7D%3B%0D%0A%0D%0Alet%20k%3A%20keyof%20T%3B%0D%0AinitState%5Bk%5D%20%3D%20'test'%20as%20any%3B%0D%0A

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we reduce empty intersections to `never` immediately upon construction. For example, instead of preserving the type `string & number` and only reducing it away in union types we now immediately reduce it to `never`.

Some examples of types that now immediately reduce to never:

```ts
type T1 = 'a' & 'b';  // never
type T2 = { a: string } & null;  // never in strict mode, null in non-strict mode
type T3 = { a: string } & undefined;  // never in strict mode, undefined in non-strict mode
type T4 = string & number;  // never
type T5 = number & object;  // never
type T6 = symbol & string;  // never
type T7 = void & string;  // never
```

In #18438 we chose to preserve empty intersection types (a) to make it easier to understand how they originated and (b) to partially preserve the ability to "tag" primtive types using enum types (e.g. `string & Tag` where `Tag` is an enum). In practice this doesn't seem to outweigh the confusion these types bring about, in particular because there are subtle behavioral differences between `never` and types like `string & number` even though they really should be the same. The differences include:

* `never` is assignable to anything whereas `string & number` is assignable only to `string` or `number`.
* `any` is not assignable to `never` but is assignable to `string & number`.
* For a type `T` with no index signatures, `T[never]` resolves to `never` whereas `T[string & number]` is an error.

These differences have no practical value. Particularly confusing is that `string & boolean` reduces to `never` (because it normalizes to `string & true | string & false`, which then reduces to `never`), whereas `string & number` sticks around and behaves differently.

Fixes #31663.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot run dt

**Comment by ahejlsberg** [comment]
RWC test run is clean. DT test run has one minor difference in `ember/test/detect-instance.ts`. Types such as `null & { foo: string }` and `undefined & { foo: string }` now reduce to `never` but previously would stick around (yet disappear in union types) and allow properties to be accessed. The new behavior is definitely more consistent.

**Comment by jack-williams** [comment]
Does this still support branding via object literals?
```ts
type Tag<T> = T & { readonly brand: unique symbol };
type TaggedNumber = Tag<number>; // not never
```

**Comment by ahejlsberg** [comment]
@jack-williams It does, except you can't brand `null` or `undefined`. You sort of could before, but only in a limited sense. The minute you put `null & { brand: xxx }` in a union type it would disappear.

**Comment by ahejlsberg** [comment]
@jack-williams The only thing that changes with this PR is *when* we reduce away empty intersections. We now do it upon construction but we used to do it only when an intersection was put in a union type. The actual patterns we reduce away remain the same.

**Comment by jack-williams** [comment]
> The only thing that changes with this PR is _when_ we reduce away empty intersections. 

Ah, that makes it clearer. Thank you!

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -9783,13 +9752,18 @@ namespace ts {
...
             else {
-                includes |= flags & TypeFlags.IncludesMask;
                 if (flags & TypeFlags.AnyOrUnknown) {
                     if (type === wildcardType) includes |= TypeFlags.IncludesWildcard;
                 }
                 else if ((strictNullChecks || !(flags & TypeFlags.Nullable)) && !contains(typeSet, type)) {
+                    if (type.flags & TypeFlags.Unit && includes & TypeFlags.Unit) {
+                        // We have seen two distinct unit types which means we should reduce to an
+                        // empty intersection. Adding TypeFlags.NonPrimitive causes that to happen.
+                        includes |= TypeFlags.NonPrimitive;
```
Comment (weswigham) [inline src/compiler/checker.ts]: Counterexample:
```ts
enum A {
  Zero = 0,
  One = 1
}

type Zeroish<T> = T & 0;
type ShouldNotBeNever = Zeroish<A.Zero> | never;
```
we handle this incorrectly in `master` already - I'm just not a fan of propagating the logical error.

Comment (fatcerberus) [reply src/compiler/checker.ts]: So is the reason for that that the code highlighted considers `A.Zero` and `0` to be distinct?  I'm just trying to ascertain where in that example above the `never` creeps in.

Comment (ahejlsberg) [reply src/compiler/checker.ts]: @weswigham I think it is debatable whether we want to consider literals and literal enum members with the same underlying value to be intersectable. But either way, it is an issue that should be covered in a separate PR. Here we're just concerned with changing *when* we reduce, not *how*.

**Comment by ahejlsberg** [comment]
@typescript-bot run dt

**Comment by ahejlsberg** [comment]
RWC and DT test runs are both clean now.

**Comment by dharnil** [comment]
Hi I wanted to check to see if this has already been resolved as I am still facing this issue.

**Pull request merged**